### ISSUE TITLE
Eval is non-deterministic across re-runs (TF32 matmul drift on H100)

### ISSUE THREAD

**Issue by LSailer** [LSailer/Master-Thesis-3D-VLA#101]
## Repro

Two runs of L1 VGGT eval, identical seed=42, identical checkpoint `step_000300000.pkl`, identical curriculum `level1_1house_1goal.json`, on the same H100:

- **Run 1** (`output/runs/r2dreamer-curriculum-l1-vggt/baseline-actent3e-4/debug/eval-pick-20ep/`, commit 9c8057d, `--episodes 20`)
- **Run 2** (`output/runs/r2dreamer-curriculum-l1-vggt/baseline-actent3e-4/debug/viz-pair-a-attempt1/`, commit d96b459, `--episodes 12 --dump_episodes 7,11`)

| Episode | Run 1 | Run 2 |
|---|---|---|
| ep0 | 500 fail | 500 fail (smoke verified all 500 actions match) |
| ep7 | 284 success, SPL 0.894 | 135 success, SPL 0.78 |
| ep11 | 500 fail (NEAR-MISS, end 0.83 m) | **57 success, SPL 0.96** |

For ep11: first 18 actions identical between runs, then diverge at step 18 (run1=`3` TURN_RIGHT, run2=`1` MOVE_FORWARD). Same observation history → different action.

## Hypothesis

TF32 default on H100 caps fp32 matmul precision at ~1e-3. cuBLAS/cuDNN kernel auto-selection produces non-deterministic reduction order. Per-matmul noise of ~1e-3 accumulates across hundreds of matmuls (VGGT backbone + JAX RSSM + actor head) over hundreds of steps, eventually flipping argmax when top-2 logits are close.

Consistent with prior knowledge: TF32 default caps fp32 parity at ~1e-3 (see VGGT-JAX parity work).

## Action selection has zero RNG

`R2DreamerAgent.act(..., training=False)` does `argmax`, no `jax.random.split`. So PRNGKey(42) is not the relevant determinism knob — bit-identical floating-point output is.

## Workaround for thesis methodology

Pair selection (success / near-miss) and per-step latent dumps **must come from the same run**. Any "pick a pair from run A, re-roll same indices in run B with instrumentation" plan is broken.

The debug-viz pipeline switched to dumping all 12 episodes in one run + selecting the pair after the fact, sidestepping this issue.

## Possible fix (if true determinism is needed later)

```python
# JAX side (modules/r2dreamer/launch/*.py entry points)
import jax
jax.config.update("jax_default_matmul_precision", "highest")

# PyTorch side (modules/vggt/feature_extractor.py)
import torch
torch.backends.cuda.matmul.allow_tf32 = False
torch.backends.cudnn.allow_tf32 = False
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
```

Plus env: `CUBLAS_WORKSPACE_CONFIG=:4096:8`. Expected slowdown: ~30–50%. Some PyTorch ops may have no deterministic kernel and will raise.

## Priority

P3. Not blocking the debug-viz work (workaround above). Worth fixing before any future eval-reproducibility claim or any "exactly N evals → exact same metrics" thesis statement.

## Affected files

- `modules/r2dreamer/launch/evaluate.py`
- `modules/vggt/feature_extractor.py`
- `scripts/debug_viz/evaluate_debug.py`
